File: tests/test_reranker_advanced.py
"""
Tests for Core/reranker_advanced.py — AdvancedReranker 5-factor scoring

Tests cover:
  - Factor calculations (semantic, position, density, answer-type, source)
  - rerank() public API
  - Weight override
  - Edge cases (empty chunks, no model)
"""

import pytest
from unittest.mock import MagicMock
import numpy as np


# ─── Fixtures ──────────────────────────────────────────────────────────────


@pytest.fixture
def mock_model():
    """Mock SentenceTransformer returning predictable embeddings."""
    model = MagicMock()

    def encode_fn(texts, **kwargs):
        results = []
        for t in texts:
            vec = np.zeros(384)
            idx = hash(t) % 384
            vec[idx] = 1.0
            results.append(vec / (np.linalg.norm(vec) + 1e-10))
        return np.array(results)

    model.encode = encode_fn
    return model


@pytest.fixture
def reranker(mock_model):
    """Create AdvancedReranker with mocked model, no CrossEncoder."""
    from Core.reranker_advanced import AdvancedReranker

    # Bypass __init__ to avoid downloading real CrossEncoder model
    r = AdvancedReranker.__new__(AdvancedReranker)
    r.model = mock_model
    r.cross_encoder = None
    return r


@pytest.fixture
def sample_chunks():
    """Small set of realistic chunks."""
    return [
        {
            "text": "Python was created by Guido van Rossum and first released in 1991. "
            "It emphasizes code readability and supports multiple programming paradigms.",
            "url": "https://en.wikipedia.org/wiki/Python",
            "source": "wikipedia",
            "title": "Python (programming language)",
        },
        {
            "text": "Subscribe to our tech newsletter for weekly Python tips and tricks!",
            "url": "https://example-blog.com/newsletter",
            "source": "blog",
            "title": "Newsletter",
        },
        {
            "text": "Step 1: Install Python from python.org. Step 2: Verify with python --version. "
            "Step 3: Set up a virtual environment.",
            "url": "https://docs.python.org/install",
            "source": "official",
            "title": "Installation Guide",
        },
        {
            "text": "A study published in IEEE found that Python adoption grew 25% in 2025, "
            "surpassing JavaScript in data science applications.",
            "url": "https://research.ieee.org/paper123",
            "source": "academic",
            "title": "Language Adoption Trends",
        },
    ]


# ─── Import ────────────────────────────────────────────────────────────────


def test_import():
    from Core.reranker_advanced import AdvancedReranker

    assert AdvancedReranker is not None


# ─── Basic rerank() ────────────────────────────────────────────────────────


def test_rerank_returns_tuple(reranker, sample_chunks):
    """rerank() should return (chunks, scores) tuple."""
    ranked, scores = reranker.rerank("What is Python?", sample_chunks, top_k=3)
    assert isinstance(ranked, list)
    assert isinstance(scores, list)
    assert len(ranked) == 3
    assert len(scores) == 3


def test_rerank_scores_sorted(reranker, sample_chunks):
    """Scores should be in descending order."""
    _, scores = reranker.rerank("How to install Python?", sample_chunks, top_k=4)
    for i in range(len(scores) - 1):
        assert scores[i] >= scores[i + 1], "Scores not sorted descending"


def test_rerank_top_k_limit(reranker, sample_chunks):
    """Should respect top_k parameter."""
    ranked, scores = reranker.rerank("Python history", sample_chunks, top_k=2)
    assert len(ranked) == 2
    assert len(scores) == 2


def test_rerank_top_k_larger_than_chunks(reranker, sample_chunks):
    """top_k > len(chunks) should return all chunks."""
    ranked, scores = reranker.rerank("Python", sample_chunks, top_k=100)
    assert len(ranked) == len(sample_chunks)


# ─── Weight Override ──────────────────────────────────────────────────────


def test_rerank_custom_weights(reranker, sample_chunks):
    """Custom weights should still produce valid results."""
    weights = {
        "semantic": 0.80,
        "position": 0.0,
        "density": 0.0,
        "answer_type": 0.10,
        "source": 0.10,
    }
    ranked, scores = reranker.rerank("What is Python?", sample_chunks, weights=weights)
    assert len(ranked) > 0
    assert all(isinstance(s, (int, float)) for s in scores)


# ─── Source Credibility ──────────────────────────────────────────────────


def test_source_credibility_ordering(reranker):
    """Official sources should rank higher than blogs."""
    from Core.reranker_advanced import AdvancedReranker

    assert AdvancedReranker.SOURCE_CREDIBILITY["official"] > AdvancedReranker.SOURCE_CREDIBILITY["blog"]
    assert AdvancedReranker.SOURCE_CREDIBILITY["academic"] > AdvancedReranker.SOURCE_CREDIBILITY["forum"]
    assert AdvancedReranker.SOURCE_CREDIBILITY["gov"] > AdvancedReranker.SOURCE_CREDIBILITY["reddit"]


# ─── Query Type Detection ─────────────────────────────────────────────────


def test_query_type_patterns_exist(reranker):
    """All expected query types should be defined."""
    expected = {
        "definition",
        "procedure",
        "comparison",
        "location",
        "temporal",
        "quantitative",
        "causal",
    }
    from Core.reranker_advanced import AdvancedReranker

    assert expected.issubset(set(AdvancedReranker.QUERY_TYPE_PATTERNS.keys()))


def test_definition_query_detected(reranker):
    """'What is X' should match definition pattern."""
    from Core.reranker_advanced import AdvancedReranker

    pattern = AdvancedReranker.QUERY_TYPE_PATTERNS["definition"]
    assert pattern.search("What is Python?")


def test_procedure_query_detected(reranker):
    """'How to ...' should match procedure pattern."""
    from Core.reranker_advanced import AdvancedReranker

    pattern = AdvancedReranker.QUERY_TYPE_PATTERNS["procedure"]
    assert pattern.search("How to install Python?")


# ─── Edge Cases ────────────────────────────────────────────────────────────


def test_empty_chunks(reranker):
    """Empty chunk list should return empty results."""
    ranked, scores = reranker.rerank("anything", [], top_k=5)
    assert ranked == []
    assert scores == []


def test_single_chunk(reranker):
    """Single chunk should be returned as-is."""
    chunk = [{"text": "The only chunk available about tropical forests."}]
    ranked, scores = reranker.rerank("tropical forests", chunk, top_k=5)
    assert len(ranked) == 1


def test_chunk_missing_optional_fields(reranker):
    """Chunks without url/source/title should still work."""
    chunks = [
        {"text": "First chunk with content about algorithms."},
        {"text": "Second chunk about data structures in computer science."},
    ]
    ranked, scores = reranker.rerank("algorithms", chunks, top_k=2)
    assert len(ranked) == 2

File: rust_output/src/test_reranker_advanced.rs
/// Tests for Core/reranker_advanced::py — AdvancedReranker 5-factor scoring
/// 
/// Tests cover:
/// - Factor calculations (semantic, position, density, answer-type, source)
/// - rerank() public API
/// - Weight override
/// - Edge cases (empty chunks, no model)

use anyhow::{Result, Context};
use std::collections::HashMap;
use std::collections::HashSet;

/// Mock SentenceTransformer returning predictable embeddings.
pub fn mock_model() -> () {
    // Mock SentenceTransformer returning predictable embeddings.
    let mut model = MagicMock();
    let encode_fn = |texts| {
        let mut results = vec![];
        for t in texts.iter() {
            let mut vec = numpy.zeros(384);
            let mut idx = (hash(t) % 384);
            vec[idx] = 1.0_f64;
            results.push((vec / (numpy.linalg.norm(vec) + 1e-10_f64)));
        }
        numpy.array(results)
    };
    model.encode = encode_fn;
    model
}

/// Create AdvancedReranker with mocked model, no CrossEncoder.
pub fn reranker(mock_model: String) -> () {
    // Create AdvancedReranker with mocked model, no CrossEncoder.
    // TODO: from Core.reranker_advanced import AdvancedReranker
    let mut r = AdvancedReranker.__new__(AdvancedReranker);
    r.model = mock_model;
    r.cross_encoder = None;
    r
}

/// Small set of realistic chunks.
pub fn sample_chunks() -> () {
    // Small set of realistic chunks.
    vec![HashMap::from([("text".to_string(), "Python was created by Guido van Rossum and first released in 1991. It emphasizes code readability and supports multiple programming paradigms.".to_string()), ("url".to_string(), "https://en::wikipedia.org/wiki/Python".to_string()), ("source".to_string(), "wikipedia".to_string()), ("title".to_string(), "Python (programming language)".to_string())]), HashMap::from([("text".to_string(), "Subscribe to our tech newsletter for weekly Python tips and tricks!".to_string()), ("url".to_string(), "https://example-blog.com/newsletter".to_string()), ("source".to_string(), "blog".to_string()), ("title".to_string(), "Newsletter".to_string())]), HashMap::from([("text".to_string(), "Step 1: Install Python from python.org. Step 2: Verify with python --version. Step 3: Set up a virtual environment.".to_string()), ("url".to_string(), "https://docs.python.org/install".to_string()), ("source".to_string(), "official".to_string()), ("title".to_string(), "Installation Guide".to_string())]), HashMap::from([("text".to_string(), "A study published in IEEE found that Python adoption grew 25% in 2025, surpassing JavaScript in data science applications.".to_string()), ("url".to_string(), "https://research.ieee.org/paper123".to_string()), ("source".to_string(), "academic".to_string()), ("title".to_string(), "Language Adoption Trends".to_string())])]
}

pub fn test_import() -> () {
    // TODO: from Core.reranker_advanced import AdvancedReranker
    assert!(AdvancedReranker.is_some());
}

/// rerank() should return (chunks, scores) tuple.
pub fn test_rerank_returns_tuple(reranker: String, sample_chunks: String) -> () {
    // rerank() should return (chunks, scores) tuple.
    let (mut ranked, mut scores) = reranker::rerank("What is Python?".to_string(), sample_chunks, /* top_k= */ 3);
    assert!(/* /* isinstance(ranked, list) */ */ true);
    assert!(/* /* isinstance(scores, list) */ */ true);
    assert!(ranked.len() == 3);
    assert!(scores.len() == 3);
}

/// Scores should be in descending order.
pub fn test_rerank_scores_sorted(reranker: String, sample_chunks: String) -> () {
    // Scores should be in descending order.
    let (_, mut scores) = reranker::rerank("How to install Python?".to_string(), sample_chunks, /* top_k= */ 4);
    for i in 0..(scores.len() - 1).iter() {
        assert!(scores[&i] >= scores[(i + 1)], "Scores not sorted descending");
    }
}

/// Should respect top_k parameter.
pub fn test_rerank_top_k_limit(reranker: String, sample_chunks: String) -> () {
    // Should respect top_k parameter.
    let (mut ranked, mut scores) = reranker::rerank("Python history".to_string(), sample_chunks, /* top_k= */ 2);
    assert!(ranked.len() == 2);
    assert!(scores.len() == 2);
}

/// top_k > len(chunks) should return all chunks.
pub fn test_rerank_top_k_larger_than_chunks(reranker: String, sample_chunks: String) -> () {
    // top_k > len(chunks) should return all chunks.
    let (mut ranked, mut scores) = reranker::rerank("Python".to_string(), sample_chunks, /* top_k= */ 100);
    assert!(ranked.len() == sample_chunks.len());
}

/// Custom weights should still produce valid results.
pub fn test_rerank_custom_weights(reranker: String, sample_chunks: String) -> () {
    // Custom weights should still produce valid results.
    let mut weights = HashMap::from([("semantic".to_string(), 0.8_f64), ("position".to_string(), 0.0_f64), ("density".to_string(), 0.0_f64), ("answer_type".to_string(), 0.1_f64), ("source".to_string(), 0.1_f64)]);
    let (mut ranked, mut scores) = reranker::rerank("What is Python?".to_string(), sample_chunks, /* weights= */ weights);
    assert!(ranked.len() > 0);
    assert!(scores.iter().map(|s| /* /* isinstance(s, (int, float) */) */ true).collect::<Vec<_>>().iter().all(|v| *v));
}

/// Official sources should rank higher than blogs.
pub fn test_source_credibility_ordering(reranker: String) -> () {
    // Official sources should rank higher than blogs.
    // TODO: from Core.reranker_advanced import AdvancedReranker
    assert!(AdvancedReranker.SOURCE_CREDIBILITY["official".to_string()] > AdvancedReranker.SOURCE_CREDIBILITY["blog".to_string()]);
    assert!(AdvancedReranker.SOURCE_CREDIBILITY["academic".to_string()] > AdvancedReranker.SOURCE_CREDIBILITY["forum".to_string()]);
    assert!(AdvancedReranker.SOURCE_CREDIBILITY["gov".to_string()] > AdvancedReranker.SOURCE_CREDIBILITY["reddit".to_string()]);
}

/// All expected query types should be defined.
pub fn test_query_type_patterns_exist(reranker: String) -> () {
    // All expected query types should be defined.
    let mut expected = HashSet::from(["definition".to_string(), "procedure".to_string(), "comparison".to_string(), "location".to_string(), "temporal".to_string(), "quantitative".to_string(), "causal".to_string()]);
    // TODO: from Core.reranker_advanced import AdvancedReranker
    assert!(expected.issubset(AdvancedReranker.QUERY_TYPE_PATTERNS.keys().into_iter().collect::<HashSet<_>>()));
}

/// 'What is X' should match definition pattern.
pub fn test_definition_query_detected(reranker: String) -> () {
    // 'What is X' should match definition pattern.
    // TODO: from Core.reranker_advanced import AdvancedReranker
    let mut pattern = AdvancedReranker.QUERY_TYPE_PATTERNS["definition".to_string()];
    assert!(pattern.search("What is Python?".to_string()));
}

/// 'How to ...' should match procedure pattern.
pub fn test_procedure_query_detected(reranker: String) -> () {
    // 'How to ...' should match procedure pattern.
    // TODO: from Core.reranker_advanced import AdvancedReranker
    let mut pattern = AdvancedReranker.QUERY_TYPE_PATTERNS["procedure".to_string()];
    assert!(pattern.search("How to install Python?".to_string()));
}

/// Empty chunk list should return empty results.
pub fn test_empty_chunks(reranker: String) -> () {
    // Empty chunk list should return empty results.
    let (mut ranked, mut scores) = reranker::rerank("anything".to_string(), vec![], /* top_k= */ 5);
    assert!(ranked == vec![]);
    assert!(scores == vec![]);
}

/// Single chunk should be returned as-is.
pub fn test_single_chunk(reranker: String) -> () {
    // Single chunk should be returned as-is.
    let mut chunk = vec![HashMap::from([("text".to_string(), "The only chunk available about tropical forests.".to_string())])];
    let (mut ranked, mut scores) = reranker::rerank("tropical forests".to_string(), chunk, /* top_k= */ 5);
    assert!(ranked.len() == 1);
}

/// Chunks without url/source/title should still work.
pub fn test_chunk_missing_optional_fields(reranker: String) -> () {
    // Chunks without url/source/title should still work.
    let mut chunks = vec![HashMap::from([("text".to_string(), "First chunk with content about algorithms.".to_string())]), HashMap::from([("text".to_string(), "Second chunk about data structures in computer science.".to_string())])];
    let (mut ranked, mut scores) = reranker::rerank("algorithms".to_string(), chunks, /* top_k= */ 2);
    assert!(ranked.len() == 2);
}
